File: 백준/Silver/2156. 포도주 시식/포도주 시식.py
n = int(input())

num = [0] * 10001
dp = [0] * 10001

for i in range(n):
    num[i] = int(input())

dp[0] = num[0]
dp[1] = num[0] + num[1]
dp[2] = max(dp[1], num[0]+num[2], num[1]+num[2])

for i in range(3, n):
    dp[i] = max(dp[i-1], dp[i-2]+num[i], dp[i-3]+num[i-1]+num[i])

print(max(dp))

File: 백준/Silver/2156. 포도주 시식/포도주 시식.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int num = Integer.parseInt(br.readLine());
        int[] numlist = new int[num];
        int[] dp = new int[num];
        for(int i = 0 ; i < num ; i++)
            numlist[i] = Integer.parseInt(br.readLine());

        if(num < 2)
            System.out.println(numlist[0]);
        else if(num < 3)
            System.out.println(numlist[0] + numlist[1]);
        else {
            dp[0] = numlist[0];
            dp[1] = Math.max(dp[0], dp[0] + numlist[1]);
            dp[2] = Math.max(dp[0] + numlist[2], numlist[1] + numlist[2]);
            dp[2] = Math.max(dp[2], dp[1]);
            for (int i = 3; i < num; i++) {
                dp[i] = Math.max(numlist[i - 1] + numlist[i] + dp[i - 3], dp[i - 2] + numlist[i]);
                dp[i] = Math.max(dp[i], dp[i - 1]);
            }

            System.out.println(dp[num-1]);
        }
    }
}
